File: app/src/main/java/wyyu/multi/cell_class/ActivityListClass.java
package wyyu.multi.cell_class;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import com.wyyu.multi.cell.ClassCellManager;
import java.util.ArrayList;
import java.util.List;
import rx.Observable;
import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.schedulers.Schedulers;
import wyyu.multi.R;
import wyyu.multi.app.Constants;
import wyyu.multi.cell_class.adapter.AdapterListClass;
import wyyu.multi.cell_class.cell.CellClassA;
import wyyu.multi.cell_class.cell.CellClassB;
import wyyu.multi.cell_class.data.DataClassA;
import wyyu.multi.cell_class.data.DataClassB;

/**
 * Created by wyyu on 2020-05-07.
 **/

public class ActivityListClass extends AppCompatActivity {

    public static void open(Context context) {
        context.startActivity(new Intent(context, ActivityListClass.class));
    }

    @BindView(R.id.class_recycler) RecyclerView recyclerView;

    private AdapterListClass<Class<?>, Object> adapter;
    private Unbinder unbinder;

    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_list_class);

        unbinder = ButterKnife.bind(this);
        initActivity();
    }

    @Override protected void onDestroy() {
        super.onDestroy();
        if (unbinder != null) {
            unbinder.unbind();
        }
    }

    private void initActivity() {
        initView();
        loadList();
    }

    private void initView() {
        adapter = new AdapterListClass<>(new ClassCellManager());

        adapter.register(DataClassA.class, new CellClassA());
        adapter.register(DataClassB.class, new CellClassB());

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(adapter);
    }

    private void loadList() {
        Observable.unsafeCreate(new Observable.OnSubscribe<List<Object>>() {
            @Override public void call(Subscriber<? super List<Object>> subscriber) {
                List<Object> objectList = new ArrayList<>();
                for (int index = 0; index < Constants.COUNT; index++) {
                    if (index % 2 == 0) {
                        objectList.add(new DataClassA(index));
                    } else {
                        objectList.add(new DataClassB(index));
                    }
                }
                subscriber.onNext(objectList);
                subscriber.onCompleted();
            }
        })
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(new Action1<List<Object>>() {
                @Override public void call(List<Object> objects) {
                    if (adapter != null) {
                        adapter.initList(objects);
                    }
                }
            }, new Action1<Throwable>() {
                @Override public void call(Throwable throwable) {

                }
            });
    }
}

File: app/src/main/java/com/wyyu/multiktest/cell_class/ActivityListClass.kt
package com.wyyu.multiktest.cell_class

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.wyyu.multi.cell.ClassCellManager
import com.wyyu.multiktest.app.Constants
import com.wyyu.multiktest.cell_class.adapter.ListClassAdapter
import com.wyyu.multiktest.cell_class.data.DataClassA
import com.wyyu.multiktest.cell_class.data.DataClassB
import com.wyyu.multiktest.cell_class.holder.HolderClassA
import com.wyyu.multiktest.cell_class.holder.HolderClassB
import com.wyyu.multiktest.cell_class.holder.HolderClassJ
import com.wyyu.multiktest.databinding.ActivityListClassBinding

/**
 * Created by wyyu on 2021/7/30.
 **/

class ActivityListClass : AppCompatActivity() {

    companion object {

        fun open(context: Context) {
            context.startActivity(Intent(context, ActivityListClass::class.java))
        }
    }

    private var binding: ActivityListClassBinding? = null
    private var adapter: ListClassAdapter<Class<out Any>, Any>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityListClassBinding.inflate(layoutInflater)
        setContentView(binding?.root)

        initActivity()
    }

    private fun initActivity() {
        initView()
        loadList()
    }

    private fun initView() {
        if (binding == null) {
            return
        }
        adapter = ListClassAdapter(ClassCellManager(), binding!!.listClassRecycler)
        adapter!!.registerItem(DataClassA::class.java, HolderClassJ())
        adapter!!.registerItem(DataClassB::class.java, HolderClassJ())

        binding?.listClassRecycler?.layoutManager = LinearLayoutManager(this)
        binding?.listClassRecycler?.adapter = adapter
    }

    private fun loadList() {
        val list = ArrayList<Any>()
        for (index in 0 until Constants.COUNT) {
            if (index % 2 == 0) {
                list.add(DataClassA(index))
            } else {
                list.add(DataClassB(index))
            }
        }
        adapter?.initList(list)
    }
}